mfr/server/handlers/render.py

```diff
@@ -12,83 +12,149 @@
-
-        # Spin off upload into non-blocking operation
-        if renderer.cache_result and settings.CACHE_ENABLED:
-            asyncio.ensure_future(
-                self.cache_provider.upload(
-                    waterbutler.core.streams.StringStream(rendition),
-                    self.cache_file_path
-                )
+                self.track_cache_miss(e)
+
+        # Ok so the file needs to be rendered.
+        # Is the file in a format that is renderable?
+        renderable = ['.pdf']
+        if not self.metadata.ext.lower() in renderable:
+            # The file is required and it's not a renderable fromat, so it
+            # needs to be converted.
+            self.convert_file()
+
+        # Perform the render and write the result to the response
+        rendition = await self.render()
+        await self.write_stream(rendition)
+
+        # Spin off upload of cached render into non-blocking operation
+        if self.render.cache_result and settings.CACHE_ENABLED:
+            self.cache_result(rendition)
+
+    async def get_cached_render(self):
+        """Fetches and returns the response from the cache"""
+        cached_render = await self.cache_provider.download(await self.cache_file_path)
+        logger.info('Cached file found; Sending downstream [{}]'.format(self.cache_file_path))
+        self.metrics.add('cache_file.result', 'hit')
+        return cached_render
+
+    def track_cache_miss(self, e):
+        """Track and log requests to the cache that are misses"""
+        assert e.code == 404, 'Non-404 DownloadError {!r}'.format(e)
+        logger.info('No cached file found; Starting render[{}]'.format(self.cache_file_path.result))
+        self.metrics.add('cache_file.result', 'miss')
+
+    @property
+    def source_stream(self):
+        """The stream to use as input. Awaitable in arder to be able
+        to pass it around without having to actually download it until it's
+        needed.
+        """
+        try:
+            return self._source_stream
+        except:
+            self._source_stream = self.provider.download()
+            return self._source_stream
+
+    @property
+    def export_url(self):
+        try:
+            return self._export_url
+        except:
+            self._export_url = self.request.uri.replace('/render?', '/export?', 1)
+            return self._export_url
+
+    def convert_file(self):
+        """The source stream is an awaitable that yields a stream. In order to
+        avoid actual download unless it's necessary. this changes the future
+        from the original waterbutler-sourced file to the future returned by
+        exporting the file.
+
+        Also change the url for the file to an export url
+        for the case that the file is not required."""
+        try:
+            map = RENDER_MAP[self.metadata.ext]
+        except KeyError:
+            map = DEFAULT_RENDER
+        self._source_stream = utils.bind_convert(
```

Comment:
make this a lambda(?) and handle the case in the property so it's not called unless needed